database/src/main/java/com/firebase/ui/database/FirebaseRecyclerAdapter.java

```diff
@@ -63,32 +62,23 @@
  *         }
  *     };
  *     recycler.setAdapter(mAdapter);
- * }
- * </pre></blockquote>
+ * </pre>
  *
- * @param <T> The Java class that maps to the type of objects stored in the Firebase location.
+ * @param <T>  The Java class that maps to the type of objects stored in the Firebase location.
  * @param <VH> The ViewHolder class that contains the Views in the layout that is shown for each object.
  */
 public abstract class FirebaseRecyclerAdapter<T, VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<VH> {
 
-    Class<T> mModelClass;
     protected int mModelLayout;
+    Class<T> mModelClass;
     Class<VH> mViewHolderClass;
     FirebaseArray mSnapshots;
 
-    /**
-     * @param modelClass Firebase will marshall the data at a location into an instance of a class that you provide
-     * @param modelLayout This is the layout used to represent a single item in the list. You will be responsible for populating an
-     *                    instance of the corresponding view with the data from an instance of modelClass.
-     * @param viewHolderClass The class that hold references to all sub-views in an instance modelLayout.
-     * @param ref        The Firebase location to watch for data changes. Can also be a slice of a location, using some
-     *                   combination of <code>limit()</code>, <code>startAt()</code>, and <code>endAt()</code>
-     */
-    public FirebaseRecyclerAdapter(Class<T> modelClass, int modelLayout, Class<VH> viewHolderClass, Query ref) {
+    FirebaseRecyclerAdapter(Class<T> modelClass, int modelLayout, Class<VH> viewHolderClass, FirebaseArray snapshots) {
```

Comment:
Make `FirebaseArray` implement an interface like `FirebaseArrayInterface`. And using interface in constructor instead of `FirebaseArray`. That will help we to implement a `has-a` relationship. Following is a example using:
```
public interface FirebaseArrayInterface {
    interface OnChangedListener {
        void onChanged(EventType type, int index, int oldIndex);

        void onFiltered();

        void onCancelled(DatabaseError databaseError);

        enum EventType {Added, Changed, Removed, Moved}
    }

    DataSnapshot getItem(int index);

    int getCount();

    void cleanup();

    void setOnChangedListener(OnChangedListener listener);
}
```

Filterable Array
```
public abstract class FirebaseFilterableArray implements FirebaseArrayInterface
        , FirebaseArrayInterface.OnChangedListener
        , Filterable {
    private FirebaseFilter mFilter;
    private List<DataSnapshot> mFiltered = new ArrayList<>();
    private FirebaseArrayInterface mSnapshots;
    private CharSequence mConstraint;
    private OnChangedListener mListener;

    public FirebaseFilterableArray(FirebaseArrayInterface snapshots) {
        mFilter = new FirebaseFilter();
        mSnapshots = snapshots;
        mSnapshots.setOnChangedListener(this);
    }

    @Override
    public int getCount() {
        return mFiltered.size();
    }

    @Override
    public DataSnapshot getItem(int index) {
        return mFiltered.get(index);
    }

    @Override
    public void cleanup() {
        mSnapshots.cleanup();
    }

    @Override
    public void setOnChangedListener(OnChangedListener listener) {
        mListener = listener;
    }

    @Override
    public void onChanged(EventType type, int index, int oldIndex) {
        if (mListener != null) {
            mListener.onChanged(type, index, oldIndex);
        }
        getFilter().filter(mConstraint);
    }

    @Override
    public void onFiltered() {
        if (mListener != null) {
            mListener.onFiltered();
        }
    }

    @Override
    public void onCancelled(DatabaseError databaseError) {
        if (mListener != null) {
            mListener.onCancelled(databaseError);
        }
    }

    @Override
    public Filter getFilter() {
        return mFilter;
    }

    private class FirebaseFilter extends Filter {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            mConstraint = constraint;

            List<DataSnapshot> result = new ArrayList<>();
            for (int i = 0; i < mSnapshots.getCount(); i++) {
                DataSnapshot item = mSnapshots.getItem(i);
                if (match(item, constraint)) {
                    result.add(item);
                }
            }

            FilterResults filterResults = new FilterResults();
            filterResults.values = result;
            return filterResults;
        }

        @SuppressWarnings("unchecked")
        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            mFiltered = (List<DataSnapshot>) results.values;
            if (mListener != null) {
                mListener.onFiltered();
            }
        }
    }

    protected abstract boolean match(DataSnapshot dataSnapshot, CharSequence constraint);
}
```

Example using
```
            mArrayFilter = new FirebaseFilterableArray(new FirebaseIndexArray(mListQuery,
                    mRootRef.child(BuildConfig.FIREBASE_LAST_MILE_ORDERS))) {
                @Override
                protected boolean match(DataSnapshot dataSnapshot, CharSequence constraint) {
                    if (constraint == null || constraint.length() == 0) {
                        return true;
                    } else {
                        LastMile lastMile = dataSnapshot.getValue(LastMile.class);
                        return lastMile.getDeliveryAddress().contains(constraint)
                                || lastMile.getRecipientName().contains(constraint);
                    }
                }
            };
            mAdapter = new FirebaseRecyclerAdapter<LastMile, LastMileRecentsViewHolder>(LastMile.class,
                    R.layout.fragment_last_mile_recents_item, LastMileRecentsViewHolder.class, mArrayFilter) {

                @Override
                protected void populateViewHolder(LastMileRecentsViewHolder viewHolder, LastMile lastMile, int position) {
                }
            };
```

Sorry, my English isn't good. :(